### ISSUE TITLE
Stateful global metrics with multi-device distribution

### ISSUE THREAD

**Issue by tstumm** [tensorflow/tensorflow#46985]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.4.1
- Are you willing to contribute it (Yes/No): Unsure

**Describe the feature and the current behavior/state.**
Assume you want to create a metric which is calculated over the full dataset (i.e. evaluating how good a model is at sorting via some regression problem). On a single GPU you could do this:
```python
import tensorflow as tf
from tensorflow.python.keras.backend import track_variable

class MyCustomGlobalMetric(tf.keras.metrics.Metric):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.y_true = tf.Variable(name='y_true', shape=tf.TensorShape(None), dtype=tf.dtypes.float32,
                                  initial_value=[0.],
                                  validate_shape=False, synchronization=tf.VariableSynchronization.ON_READ,
                                  aggregation=tf.VariableAggregation.ONLY_FIRST_REPLICA)
        self.y_pred = tf.Variable(name='y_pred', shape=tf.TensorShape(None), dtype=tf.dtypes.float32,
                                  initial_value=[0.],
                                  validate_shape=False, synchronization=tf.VariableSynchronization.ON_READ,
                                  aggregation=tf.VariableAggregation.ONLY_FIRST_REPLICA)
        self._non_trainable_weights.append(self.y_true)
        self._non_trainable_weights.append(self.y_pred)
        track_variable(self.y_true)
        track_variable(self.y_pred)
        self.was_called = self.add_weight('was_called', initializer=tf.keras.initializers.zeros, dtype=tf.dtypes.uint8,
                                          synchronization=tf.VariableSynchronization.ON_READ,
                                          aggregation=tf.VariableAggregation.ONLY_FIRST_REPLICA)

    def update_state(self, y_true, y_pred, **kwargs):
        y_true = K.flatten(K.cast(y_true, self.y_true.dtype))
        y_pred = K.flatten(K.cast(y_pred, self.y_true.dtype))
        a = self.y_true.assign(tf.concat([self.y_true, y_true], axis=0))
        b = self.y_pred.assign(tf.concat([self.y_pred, y_pred], axis=0))
        c = self.was_called.assign(1)
        return tf.group(a, b, c)

    def result(self):
        return tf.cond(self.was_called >= 1, lambda: some_custom_metric(self.y_true[1:], self.y_pred[1:]), lambda: 0.)

    def reset_states(self):
        super().reset_states()
        self.y_true.assign(tf.constant([0.]))
        self.y_pred.assign(tf.constant([0.]))
        self.was_called.assign(0)
```
This will store all predictions and labels and calculates `some_custom_metric` for them. But this fails to calculate the correct metric when instantiated and used in a MirroredStrategy scope. Growing the variable shape is a hacky workaround to get this running, but there is no workaround in the first way. None of the `VariableAggregation` options actually make sense when you want to calculate the metric in a global sense on multiple devices.

I'd like to have the possibility to implement a stateful metric which allows me to store all predictions and labels, such that I can calculate some arbitrary metric over them, which should be consistent and correct even if I'm using a MirroredStrategy.

**Will this change the current api? How?**

**Who will benefit with this feature?**

**Any Other info.**

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
